dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIManagedObject.java

```diff
@@ -103,4 +122,90 @@ public String toString() {
         String released = this.creationalContext == null ? " (RELEASED)" : "";
         return "CDIManagedObject: " + this.identity + released;
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws CDIException
+     */
+    @Override
+    public T inject(ReferenceContext referenceContext) throws ManagedObjectException {
+        T instance = getObject();
+        if (referenceContext == null) {
+            instance = cdiInjection(instance);
+        } else {
+            com.ibm.wsspi.injectionengine.InjectionTarget[] injectionTargets;
+            try {
+                injectionTargets = referenceContext.getInjectionTargets(instance.getClass());
+            } catch (InjectionException e) {
+                throw new ManagedObjectException(e);
+            }
+
+            InjectionTargetContext injectionContext = new InjectionTargetContext() {
+                @Override
+                public <S> S getInjectionTargetContextData(Class<S> data) {
+                    return getContextData(data);
+                }
+            };
+            instance = inject(injectionTargets, injectionContext);
+        }
+
+        return instance;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public T inject(com.ibm.wsspi.injectionengine.InjectionTarget[] targets, InjectionTargetContext injectionContext) throws ManagedObjectException {
+        T instance = getObject();
+        synchronized (this) { //we really don't want inject being called twice at the same time
+
+            try {
+                this.currentInjectionTargets = new HashSet<>();
+                this.currentInjectionContext = injectionContext;
+                this.websphereInjectionServices.registerInjectionTargetListener(this);
+
+                for (com.ibm.wsspi.injectionengine.InjectionTarget it : targets) {
+                    this.currentInjectionTargets.add(it);
+                }
+
+                //use Weld to perform injection
+                instance = cdiInjection(instance);
+
+                for (com.ibm.wsspi.injectionengine.InjectionTarget it : this.currentInjectionTargets) {
+                    if (it.getInjectionBinding().getAnnotationType() != Inject.class) {
+                        try {
+                            it.inject(instance, injectionContext);
+                        } catch (InjectionException e) {
+                            throw new ManagedObjectException(e);
+                        }
+                    }
+                }
+            } finally {
+                this.websphereInjectionServices.deregisterInjectionTargetListener(this);
+                this.currentInjectionContext = null;
+                this.currentInjectionTargets = null;
+            }
+
+        }
+
+        return instance;
+    }
+
+    T cdiInjection(T instance) {
```

Comment:
does this really need to be a separate method? should the instance param be annotated @sensitive ?

Reply:
I say keep it, it's called by two methods and a nice spot for entry/exit trace.